src/coreclr/vm/frames.cpp

```diff
@@ -929,33 +930,62 @@ GCFrame::GCFrame(Thread *pThread, OBJECTREF *pObjRefs, UINT numObjRefs, BOOL may
 
 #endif // USE_CHECKED_OBJECTREFS
 
+#ifdef _DEBUG
+    m_Next          = NULL;
+    m_pCurThread    = NULL;
+#endif // _DEBUG
+
     m_pObjRefs      = pObjRefs;
     m_numObjRefs    = numObjRefs;
-    m_pCurThread    = pThread;
     m_MaybeInterior = maybeInterior;
 
+    Push(pThread);
+}
+
+GCFrame::~GCFrame()
+{
+    WRAPPER_NO_CONTRACT;
+    Pop();
+}
+
+void GCFrame::Push(Thread* pThread)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+        PRECONDITION(pThread != NULL);
+        PRECONDITION(m_Next == NULL);
+        PRECONDITION(m_pCurThread == NULL);
+    }
+    CONTRACTL_END;
+
     // Push the GC frame to the per-thread list
     m_Next = pThread->GetGCFrame();
+    m_pCurThread = pThread;
 
     // GetOsPageSize() is used to relax the assert for cases where two Frames are
     // declared in the same source function. We cannot predict the order
-    // in which the C compiler will lay them out in the stack frame.
+    // in which the compiler will lay them out in the stack frame.
     // So GetOsPageSize() is a guess of the maximum stack frame size of any method
-    // with multiple Frames in mscorwks.dll
+    // with multiple GCFrames in coreclr.dll
     _ASSERTE(((m_Next == NULL) ||
               (PBYTE(m_Next) + (2 * GetOsPageSize())) > PBYTE(this)) &&
              "Pushing a GCFrame out of order ?");
 
     pThread->SetGCFrame(this);
 }
 
-GCFrame::~GCFrame()
+void GCFrame::Pop()
 {
     CONTRACTL
     {
         NOTHROW;
         GC_NOTRIGGER;
         MODE_ANY;
+        PRECONDITION(m_Next != NULL);
+        PRECONDITION(m_pCurThread != NULL);
     }
     CONTRACTL_END;
 
```

Comment:
We do not need any of the GC mode switching when this is called from FCall. Move the GC move switching to the destructor?